### PR TITLE
Add the ability to implement and test custom advanced visibility stores

### PR DESCRIPTION
**Pull request by aromanovich**
**What changed?**

Two things:
* `common/persistence/tests/visibility_persistence_suite_test.go` renamed to `visibility_persistence_suite.go`, so that its `VisibilityPersistenceSuite` could be imported and run by the third-party package;
* `VisibilityStoreFactory` interface introduced. The server now accepts `customVisibilityStoreFactory` and invokes it if visibility store configuration points to `CustomDataStoreConfig`.

**Why?**

To be have an ability to implement and test custom advanced visibility stores.

**How did you test it?**
Ran Temporal tests.
Also built and ran a Temporal server with `temporal.WithCustomVisibilityStoreFactory(mypackage.NewMyVisibilityStoreFactory())` option passed.

**Potential risks**
None.

**Is hotfix candidate?**
No.

### CODE REVIEW

**Comment by aromanovich** [comment]
@rodrigozhou, does this feature seem acceptable in principle? If yes, I would rebase and resolve conflicts so we could proceed with the review.

**Comment by rodrigozhou** [comment]
@aromanovich Yes, please. This feature sounds nice!

**Comment by aromanovich** [comment]
@rodrigozhou, done.

I would greatly appreciate if you could also take a look at https://github.com/temporalio/temporal/pull/5118 — it's a bit related. Merging them would make the life of https://github.com/yandex/temporal-over-ydb a lot easier :)

**Comment by aromanovich** [comment]
Gentle bump. Just in case this got lost 🙂

File: common/persistence/visibility/factory.go [hunk 1]
```diff
@@ -282,6 +297,15 @@ func newVisibilityStoreFromDataStoreConfig(
...
 		)
+	} else if dsConfig.CustomDataStoreConfig != nil {
+		visStore, err = customVisibilityStoreFactory.NewVisibilityStore(
+			*dsConfig.CustomDataStoreConfig,
+			persistenceResolver,
+			logger,
+			metricsHandler,
+		)
+	} else {
+		return nil, nil
```
Comment (rodrigozhou) [inline common/persistence/visibility/factory.go:308]: `visStore` and `err` must be `nil` if it didn't enter any of those branches, so the last return should work.

**Review by rodrigozhou — CHANGES_REQUESTED** [review]
LGTM, just added a nit comment.

**Review by rodrigozhou — APPROVED** [review]

**Comment by rodrigozhou** [comment]
@aromanovich 
Approved, but it has conflicts with main branch again. Can you rebase it so I can merge please?

**Comment by aromanovich** [comment]
@rodrigozhou Done. Semgrep is not passing, but I'm quite sure it's not due to my rebase :)

**Pull request merged**